org.eclipse.buildship.ui/src/main/java/org/eclipse/buildship/ui/view/execution/UpdateExecutionPageJob.java

```diff
@@ -12,75 +12,76 @@
 package org.eclipse.buildship.ui.view.execution;
 
 import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentMap;
 
-import org.gradle.tooling.events.OperationDescriptor;
+import org.gradle.tooling.events.FailureResult;
+import org.gradle.tooling.events.FinishEvent;
+import org.gradle.tooling.events.task.TaskOperationDescriptor;
 
 import com.google.common.base.Preconditions;
-import com.google.common.collect.Maps;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.MapMaker;
 
 import org.eclipse.core.runtime.IProgressMonitor;
 import org.eclipse.core.runtime.IStatus;
 import org.eclipse.core.runtime.Status;
 import org.eclipse.core.runtime.jobs.Job;
-import org.eclipse.jface.viewers.AbstractTreeViewer;
 import org.eclipse.jface.viewers.TreeViewer;
 import org.eclipse.swt.widgets.Display;
 import org.eclipse.ui.PlatformUI;
 
 /**
- * Updates the duration of the registered {@link OperationItem} instances in the {@link ExecutionsView} in regular intervals.
+ * Updates the duration of the registered {@link OperationItem} instances in the
+ * {@link ExecutionsView} in regular intervals.
  */
 public final class UpdateExecutionPageJob extends Job {
 
     private static final int REPEAT_DELAY = 100;
 
-    private final Map<OperationDescriptor, OperationItem> operationItems;
-    private volatile boolean running;
     private final TreeViewer treeViewer;
+    // if the value is true then the operation can be removed
+    private final ConcurrentMap<OperationItem, Boolean> activeItems;
```

Comment:
I still feel like this could be simpler. We could have a Queue of events and then just forward those events to the UI element. The UI element would then react to:

- start events by adding them to the "active operations"
- finish events by removing them from the "active operations"
- timer by updating all active event's time